enhancements/microshift/microshift-updateability-ostree.md

```diff
@@ -0,0 +1,774 @@
+
+To go in greater detail why rollback is supported and downgrade is not:
+- Rollback performed due to unhealthy (red) boot
+  - Red scripts will persist a "restore" action to perform on next boot.
+  - System is booted into older ostree commit with older MicroShift
+  - MicroShift pre-run procedure runs:
+    - Restore is performed (version in metadata is the same as one in the binary)
+    - Data migration procedure compares version in metadata and binary
+    - Data migration isn't performed, because versions match.
+- Downgrade would follow a healthy (green) boot
+  - Green scripts will persist a "backup" action to perform on next boot
+  - System is booted into older ostree commit with older MicroShift
+  - MicroShift pre-run procedure runs:
+    - Backup is performed (version in metadata is newer than a binary)
+    - Data migration procedure compares version in metadata and binary
+    - Data migration refuses to proceed because `backup version` > `binary version`.
+
+Decision to perform or refuse a data migration
+to schema compatible with newly loaded MicroShift version
+will be based on following facts:
+- version persisted in MicroShift's data dir (version that created/successfully ran using the data),
+  also referred to as (version) metadata
+- version of currently installed MicroShift binary
+- embedded in MicroShift binary list of blocked "from" versions
+
+A general flow will have following form:
+1. If persisted version is missing, assume 4.13.
+1. If version of `microshift` binary is older than version in metadata, **refuse to start MicroShift**.
+1. If persisted version is on a list of blocked version migrations, **refuse to start MicroShift**.
+1. If binary is the same version as persisted in metadata, **no need for a data migration**.
+1. Otherwise upgrade is allowed and data migration will be performed.
+
+### Data migration
+
+Data migration shall include:
+- Storage version - update of Kubernetes objects by reading Resource in older version and writing newer version
+  - Leveraging existing [Kube Storage Version Migrator](https://github.com/openshift/kubernetes-kube-storage-version-migrator) and [its Operator](https://github.com/openshift/cluster-kube-storage-version-migrator-operator).
+- etcd schema - comparing v2 and v3, most likely if/when etcd v4 is released
+  - etcd project documents how to migrate from v2 to v3, but we'll also ask OpenShift etcd team for guidance.
+
+### Open Questions [optional]
+
+- Do we want to persist ostree commit ref in the metadata?
+  - Can we make some part of the process safer, more robust, by comparing previous and current commit ref?
+  - Are we good with having previous boot MicroShift version persisted?
+
+- If green/red info is not persisted, and it's not a first boot (/var/lib/microshift exists), what should we do?
+  - Compare versions (metadata vs binary)
+    - Same - start cluster
+    - Different - refuse to start? Try to migrate (but create backup before)?
```

Comment:
If we're going to wait to write that metadata until MicroShift starts successfully, what do we do in this case if the metadata does not exist?

Reply:
Metadata is created only when MicroShift is healthy (regardless if it's done by healthcheck or green script), so missing metadata == red boot -> "restore".
If we have backup to restore from - no problem.
If we don't - see open question about "restore action but no backup"
Sounds like bug scenario? We need to watch out for this when resolving "restore action but no backup" so we don't end up in this kind of situation

Reply:
Also it might mean that we're upgrading from 4.13 (data dir exists, no backup, no metadata, no action to do)